### PR TITLE
Remove `redis-py-cluster` dependency

### PR DESCRIPTION
**Pull request by anonrig**
- removes `redis-py-cluster` dependency
- removes redis v3 support since it doesn't support RedisCluster
- updates GitHub workflows
- updates Mypy and rb

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -1,8 +1,8 @@
 black==22.6.0
 flake8==5.0.4
-mypy==0.971
+mypy==1.8
 pytest>=7.2.0
 sphinx-autodoc-typehints[type-comment]>=1.19.3
 sphinx-rtd-theme
 rb>=1.9.0
-redis-py-cluster>=2.1.0
+redis>=5.0.1
```
Comment (anonrig) [inline requirements-dev.txt:8]: Should I remove Redis version, and keep it as a requirement for installation, or just include it with the package? 

I'm +1 on shipping this with Redis. lmk wyt

**Comment by mitsuhiko** [comment]
The risk here is that we're using redis everywhere, and there are potential subtle behavior changes in all of this. An upgrade like this would need to bake on S4S or elsewhere for a significant time (multiple days) to feel confident we're not running into unexpected differences. The constraint issue you run into is not caused by `sentry-redis-tools` from what I can tell but something is hard pinning on `3.4.1`.

**Comment by untitaker** [comment]
@mitsuhiko 

> The risk here is that we're using redis everywhere, and there are potential subtle behavior changes in all of this

This PR does not implicitly migrate all apps to `redis`. sentry-redis-tools supports both `redis` and `redis-py-cluster` for talking to clusters, and this PR removes support for `redis-py-cluster`. The diff looks larger than it is.

Snuba already uses `redis` for talking to clusters, hence this library had to be agnostic over both cluster impls.

I think we should first get rid of redis-py-cluster in sentry. Getting rid of redis-py-cluster in sentry can be done more gradually, and then we can land a PR like this.

Same with rb, really.

**Comment by untitaker** [comment]
> We also can't update Redis in sentry repository because of sentry-redis-tools limitations:

we can just remove the version constraint in sentry-redis-tools if it helps, but i am not entirely sure that it does. `>=5.0.0` should not prevent you from upgrading, but maybe it forces you to upgrade further than you'd like?

File: .python-version [hunk 2]
```diff
@@ -1 +1 @@
-3.8.16
+3.11.6
```
Comment (untitaker) [inline .python-version:1]: snuba is on python 3.10, will fix soon but we can't go that far. I would also split out this change tbh

Comment (anonrig) [reply .python-version]: I've several questions: Why can't we go far? Why does this version needs to be in-sync with Snuba? Sentry repository relies on 3.11 right now.

**Review by loewenheim — APPROVED** [review]

File: .python-version (hunk 2 above)
Comment (asottile-sentry) [reply .python-version]: the main issue being we can't block upgrades of an internal library in our own services.  libraries need to support the lowest common denominator internally which is currently 3.10

Comment (anonrig) [reply .python-version]: Makes sense. I'll update it in a follow up pull request.

**Comment by mitsuhiko** [comment]
> I think we should first get rid of redis-py-cluster in sentry. Getting rid of redis-py-cluster in sentry can be done more gradually, and then we can land a PR like this.

Basically my point. Before we can remove `redis-py-cluster` here we need to stop using it in sentry :)

**Comment by anonrig** [comment]
> > I think we should first get rid of redis-py-cluster in sentry. Getting rid of redis-py-cluster in sentry can be done more gradually, and then we can land a PR like this.
> 
> Basically my point. Before we can remove `redis-py-cluster` here we need to stop using it in sentry :)

We can't stop using `redis-py-cluster` because `RedisCluster` class is implemented in Redis v4, and because of `redis-py-cluster` in this repository, which has a requirement of `<= v4`, we can't update it there. So, we have to do it here first.

**Pull request closed without merging**